### ISSUE TITLE
Behavior with controls, particularly non-native controls, overlap

### ISSUE THREAD

**Issue by gkatsev** [w3c/webvtt#503]
What should be the expected behavior of cues when controls or obscure the cues?

According to the spec, cue rendering should be re-done when the native controls are shown (steps 4 and 5 of [the Processing Model](https://w3c.github.io/webvtt/#processing-model)). There are two corresponding tests on WPT
https://github.com/web-platform-tests/wpt/blob/master/webvtt/rendering/cues-with-video/processing-model/enable_controls_reposition.html
https://github.com/web-platform-tests/wpt/blob/master/webvtt/rendering/cues-with-video/processing-model/disable_controls_reposition.html

This leads to two questions:
1. The browsers have different behavior around this. Should the spec define a consistent behavior?
2. A lot of video players nowadays do not use native controls, but try to rely on native caption support. How can we better support this use case?

For 1,
Here are the current behaviors in Chrome, Firefox, and Safari:
- Chrome will move a cue up to not overlap with the control bar and then leave the cue there for the remaining duration of the cue.
- Firefox will move the cue up and leave it there unless another cue is rendered, in which case the moved up cue will move back down.
- Safari moves the cues up and then immediately back down when the controls collapse.

In addition, it can lead to interesting bugs, where due to Chrome's taller controls section, cues may end up getting re-ordered which can confuse the viewer since the cues are reversed from the expected order. https://bugs.chromium.org/p/chromium/issues/detail?id=1141592#c_ts1630084590

For 2,
Since WebVTT limits which CSS properties are allowed, it isn't possible to target `::cue` with something like `bottom: 3em` or a `transitionY(-3em)`. However, Chrome and Safari provide a pseudo-element for the text track display area that could have a transform applied to it when the controls are active. For example:
```css
video::-webkit-media-text-track-display {
  transform: transitionY(-3em);
}
```
Firefox does not have such a mechanism. Though, even this isn't ideal because if there are cues at the top of the video it could push them out of view.

The only reliable cross browser way of moving cues so that they aren't obstructed by the control bar is to modify their `line` property. While this works easily for simple cues, it could get complicated for non-simple cues, like those that are positioned or are inside of regions.

Otherwise, you either accept that captions will be obstructed by the controls, or no longer use native rendering of captions. Both options aren't ideal.

**Comment by nigelmegitt** [comment]
+1 this is a real world issue, for captions in any format. Also worth noting is that controls nowadays are not universally positioned at the bottom of the screen, but are often large and in the vertical centre of the video area.

**Comment by gkatsev** [comment]
There's also a related concern where there are players that cover the entire display area in a semi-transparent black color to increase contrast for the controls etc., which reduces the text contrast and makes it harder to read.

**Comment by nigelmegitt** [comment]
Yes, spatially, this is a 3 dimensional problem. x and y are for horizontal and vertical positioning to avoid overlap, and z is for layering. Who defines the layer model of the controls vs the captions? Also _other_ UX that players might want to add, that go beyond controls and into more complex product features, like onward journeys, watch lists, cast information etc.

**Comment by gkatsev** [comment]
Thinking about the chrome bug, I'm not sure how this behavior can be clarified in the spec to improve user experience without rewriting the collision avoidance from scratch, which is probably not an option.

**Comment by silviapfeiffer** [comment]
I thought we had specified that the rendering area for captions changes
when controls are shown, so the Safari behavior seems most conformant. All
calculations are still the same except with a smaller rendering area, the
line heights are reduced.

I did not check the spec, but that's what I remember.

Hope that's helpful.

**Comment by nigelmegitt** [comment]
> I thought we had specified that the rendering area for captions changes when controls are shown

I haven't checked either, but if we _did_ specify that, then it wouldn't work with the current fashion to position controls in the vertical centre. It may be time to revisit this and come up with a more general solution.

**Comment by gkatsev** [comment]
The spec says that when controls are shown the areas that controls are covering are made into transparent css boxes so that when the cue css boxes are created the cue boxes could be placed where the controls won't be.
Theoretically, it should help with vertical centered controls. Though, this will only help with native captions and native controls, and still leaves us open to the bug where cues are re-ordered.

While the spec also says that whenever any of the cue settings change that the processing model algorithm for rendering cues should be re-run, it doesn't necessarily say that it must be run whenever the user agent has started or stopped showing their user interface.

At first, I thought that both should be discussed in one issue for ease of use, but I think it may be best to split this issue into two for greater clarity and more focused discussion. Thoughts?
- how to handle native captions with non-native controls and other overlapping content
- how to improve user experience when cues are repositioned as part of controls showing

**Comment by gkatsev** [comment]
I spent some time thinking about this, I wonder if we can adopt a new CSS property something like `controls-mask` and the values can be something similar to [`clip-path`](https://developer.mozilla.org/en-US/docs/Web/CSS/clip-path). Though, instead of just a single value, it could be a list for multiple shapes. Originally, I was thinking that it could be a xywh value like the [spacial media fragments](https://www.w3.org/TR/media-frags/#naming-space) but I think the clip-path direction is more flexible.

**Comment by silviapfeiffer** [comment]
Could you explain in simpler words what the exact problem is that you're
trying to solve? I read it all, but it seems to be that The speed already
addresses the mentioned problem, so I must not really understand the
problem.

Thanks,
Silvia.

**Comment by gkatsev** [comment]
Thanks, Silvia, it's worth recapping here in simpler terms. I think the TPAC conversation helped crystallize the root issue as well as a potential direction.

The core issue I want addressed is being able to **make sure that captions are unobscured by overlays**. My initial request was for native captions and custom controls, but at TPAC, we realized we probably want the reverse as well: custom captions with native controls.

While the core of this is for player controls, it could be helpful for any sort of overlay. At TPAC, one idea we had is that the video element should expose controls bounding box. This box will represent either the native controls or be overridden by users with the custom control bounding box.

Ultimately, this feature request will likely need to land in HTML itself, with potentially some changes in WebVTT to accommodate it.

@eric-carlson @jernoble @alastor0325 would love your feedback on this. Thanks!

@chrisn are you able to help figure out who we can ask to look at this at Chrome? Thanks!

**Comment by alastor0325** [comment]
> The browsers have different behavior around this. Should the spec define a consistent behavior?

It definitely sounds good if we can define a consistent behavior across browsers, which is something we would be glad to follow.

> A lot of video players nowadays do not use native controls, but try to rely on native caption support. How can we better support this use case?

If this is the case, allowing video element to know the size of customized control sounds like a good idea. Is it like using `DOMRect` to define a bounding box? Does the JS player have to update that bounding box correctly whenever their own control appears or disappears? If video gets resized, does the bounding box get resized automatically as well?

I also notice that there are also some websites using their own controller and rendering for cues. Eg. Youtube. If we can make a consensus to the consistent behavior, does that mean we should also ask those kinds of JS players to follow the behavior as well? CCed @gijsk, our video control expert, to see if he has any thought as well.

**Comment by gijsk** [comment]
> CCed @gijsk, our video control expert, to see if he has any thought as well.

I am duly flattered but I suspect @mikeconley and/or some of the folks working on our picture in picture work are a better bet here.

**Comment by mikeconley** [comment]
Hi all,

> I also notice that there are also some websites using their own controller and rendering for cues. Eg. Youtube. If we can make a consensus to the consistent behavior, does that mean we should also ask those kinds of JS players to follow the behavior as well?

From our experience adding subtitle and caption support for Firefox's Picture-in-Picture feature, it seems that most major video-playing websites (at least those considered major in North America) use their own custom rendering for cues. Here is the list of sites that we have custom wrappers for to support captions and subtitles:

* BBC iPlayer
* DailyMotion
* Disney+
* HBOMax
* Hotstar
* Hulu
* Netflix
* Piped
* Amazon Prime Video
* Tubi
* Voot
* The Washington Post
* YouTube

It would be truly lovely if we could incentivize them to use WebVTT instead so that we can stop maintaining these site-specific wrappers - although (based on my experience) I would predict non-uniform success in contacting these sites to suggest this change _unless_ we can make a strong case that it reduces their technical burden.

I'm no web-spec designer, but I do worry about the resize / move case of the DOMRect idea: a site would presumably have to constantly update that DOMRect in the event that their controls change for some reason (even if it's just the size of the video changing). Perhaps instead, the site could indicate which _elements_ constitute the controls, and the browser could take on the responsibility of tracking their size and position over time? If we're going to have success selling this idea to site authors in the wild like the ones I listed, I suspect we have to make a case that they can offload work to the browser, rather than take on more work to manage DOMRects.

**Comment by nigelmegitt** [comment]
@mikeconley From a BBC point of view, there is no way we will be using WebVTT for captions and subtitles - it does not meet our requirements in multiple ways, some of which are fixable, others architectural and unfixable. 

An alternative would be to provide support for client code to provide DOM fragments for subtitles and captions, to be displayed at specified times, to keep the UA's concerns separate from the content provider's. Safari already has a tech preview experimental version of this (CC @eric-carlson who can describe it better) that can be used as a model. There's still a question about how to resolve positioning clashes with controls, but this can at least form part of the solution, I think.

For those folk for whom WebVTT does meet their requirements, why not render the WebVTT into those same DOM fragments for presentation? I suspect that's what happens in the shadow DOM in most implementations already, anyway.

**Comment by nigelmegitt** [comment]
We need to standardise both the document format interface and the right API surface.

>who cares?

It's important for people implementing tooling and procuring products and services, so they have the right level of interoperability and expectation. One of the huge problems in the area of subtitles and captions is the proliferation of document formats, be they standards or not, as over time various newcomers have looked at the problem, underestimated the complexity and generated a new solution that both has its own problems and in some cases is a problem just by its mere existence (that's not a dig at any particular individual format, and I do recognise that in some cases there are genuine attempts to solve problems).

Looked at holistically, subtitle and caption document exchange standards are important for reducing the entry costs for folk wanting to make their media accessible.

But by tying those standards too tightly to the API, we've got ourselves in a pickle.

**Comment by gkatsev** [comment]
> a site would presumably have to constantly update that DOMRect in the event that their controls change for some reason (even if it's just the size of the video changing).

Yes, a site would need to update this, but presumably it would only happen when the controls are shown and when they are hidden.

> Perhaps instead, the site could indicate which _elements_ constitute the controls

This is a reasonable alternative, but only solves the custom controls with native captions problem and doesn't help with less popular native controls and custom captions setups.

I think the question, right now, is more would some kind of mechanism for this be reasonable to add to browsers? The specifics can be ironed out as it progresses through the various WGs.

> I have wondered for quite a while whether we standardized the wrong interface with VTT and TTML. _If_ all sites are going to implement captions using Javascript, then the important interface is not the document format those scripts interpret (who cares? the script and format merely need to match, and both are downloaded), but it is the API surface that the script can work on. Should we pay more attention to that?

I do think that a built-in format is still important to have as a base layer. However, having the right primitives in place for building upon so that not everything needs to get build from scratch would be great. For example, one missing item is [`removeTextTrack`](https://github.com/whatwg/html/issues/1921).

[Webkit's proposal](https://github.com/WebKit/explainers/blob/main/texttracks/README.md), that @nigelmegitt mentioned above, seems like the logical next step to expanding these primitives.

Ultimately, this specific issue isn't really a WebVTT issue. It only began here because I encountered it in the context of WebVTT, and I wanted some more thoughts/buy in from vendors on it before trying to push it through the Media WG and eventually the HTML spec.
